### DISCUSSION TITLE
Can i use as not-module

### DISCUSSION THREAD

**Discussion by MixtarMaster** [a2br/ecoledirecte.js#150] [Q&A] [answered]
Hello, how you all doing?

I'm new to node so this might be a stupid question. I'm trying to use this library to make a sort of bot which will send me my daily homework and/or what was done during the lesson. I re-use the code in the example, making me have to "compile?" my code as a "module" (to use import { ... } from .. and await, which i'm not sure i understand the use of either but the code won't work without it). I went around this "module" issue by using import ... from for every library i need. But now when i try to put my code in a function (to use it in loops or scheduling execution ....) it just stop working when it's as a function.

Please help!!! (don't hesitate to ask for any more info or bits of code)

**Comment by a2br** [comment, marked as answer]
# Edit
Pardon, j'avais mal compris. `await` n'est normalement pas disponible au premier niveau, il faut l'utiliser dans une fonction asynchrone:
```ts
async function thisIsAnAsyncFunction() {
  // Here, you can wait for promises with 'await'
}
```
Tu n'as pas besoin de convertir ton projet vers les modules, même si c'est plus pratique. Si ça ne marche pas dans les fonctions, c'est peut-être parce que tu n'as pas utilisé `async`. Si tu veux revenir à `CommonJS` (avec require et sans les await de premier niveau), tu peux le faire. En général, on fait ça pour utiliser await en CommonJS au premier niveau:
```ts
(async () => {
  // Here is the code; it's actually defining an async function and then executing it.
  // If you use this, don't forget to declare your imports outside, at the top of the file
})()
```
N'hésite pas à me partager ton code, je t'aiderais avec plaisir!
# Réponse originale

Salut !

Tu peux parler français ici, le reste est juste en anglais par convention. ;)
Si j'ai bien compris, tu cherches à utiliser le module. Puisque tu est nouveau sur Node.js et que tu n'as peut-être pas entendu parler d'[npm](https://npmjs.com/), peut-être que tu devrais lire ce [guide explicatif](https://nodesource.com/blog/an-absolute-beginners-guide-to-using-npm/).

La version courte :
1. D'abord, va dans le répertoire de ton projet et fais `npm init -y`. (le `-y` est pour qu'il arrête de te poser des questions). Je pense que tu l'as déjà fait si tu as déjà, par exemple, installé discord.js.
2. Maintenant, fais juste `npm install ecoledirecte.js`, et c'est bon! Pas besoin de compiler quoi que ce soit, c'est déjà fait. Si tu utilises TypeScript, tu devrais peut-être aussi faire `npm install -D ecoledirecte-api-types`.

Pour rappel, les imports se font en haut des fichiers. Ca peut se faire comme ça : 
```ts
const Ed = require("ecoledirecte.js")
```
Ou, sur les versions plus récentes ou TS, comme ça : 
```ts
import * as Ed from "ecoledirecte.js"

const session = new Ed.Session("identifiant", "motdepasse");
// etc...
```
Tu as parlé d'`await`, qui est lié à la notion de promesses. C'est un concept parfois un peu dur à comprendre, mais c'est tout simple. Tu peux en lire plus [ici](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise). C'est juste la *promesse* qu'une valeur va arriver, et mettre `await` devant une valeur promise fait le programme attendre cette valeur. On utilise les promesses pour les trucs qui prennent un bout de temps en général, comme un requête aux serveurs d'EcoleDirecte.

Dis-moi si t'as compris! Et n'hésite pas à laisser une étoile sur le repo s'il t'est utile :)

**Reply by MixtarMaster** [reply]
Merci! je pense pas avoir été clair non plus... j'avais en réalité plusieurs question: une importante puis d'autres moins importante (pour mieux comprendre)

La réelle question: lorsque j'execute le code tel quel, tout vas bien. Mais lorsque le même code je le met:
`
function theCode(){
//ici le code d'origine qui fonction
}`

au niveau de la ligne d'import, ça me sort l'erreur: 
`file:///C:/Users/PaulV/Desktop/wepAppTaff/app.js:21
import {Session} from "ecoledirecte.js"; 
SyntaxError: Unexpected token '{'`

mon "problème" avec les await c'est que pour résoudre mon problème de la fonction, je le mettais en pas module ("commonJS"?) mais alors les "await" me sortais une erreur. je vais voir si je peux résoudre le problème avec les `async function` pour rester en commonJS. Après j'aimerais bien savoir si c'est possible en restant en module

TL;DR:

`//code autre
import {Session} from "ecoledirecte.js";
//autre code`
=> pas d'erreur
`function inAFunction(){
//code autre
import {Session} from "ecoledirecte.js";
//autre code
}`
=> erreur

**Reply by MixtarMaster** [reply]
UPDATE:

Ca marche parfaitement avec `async function` en commonJS (donc pas module) merci beaucoup !

**Reply by MixtarMaster** [reply]
Aussi:

J'ai pas bien compris tout autour de git: comment je fais pour rendre mon code public (je me dis que ce que j'ai fais peut etre utile à d'autres)

**Reply by a2br** [reply]
Ah, content que ça marche. Oui, les imports se font tout le temps en *haut* du fichier. Git est un logiciel de source control, pour ce qui est de le rendre public c'est plus GitHub. En général, je fais quelque chose comme
```sh
# Crée le repo GitHub en ligne, totalement vide, sans LICENSE ou README automatique.
# Il devrait s'afficher à peu près les mêmes instructions.

# Init repo git
$ git init

# Sélectionne tous les fichiers. Tu peux en désélectionner par défaut avec un fichier .gitignore (comme les fichiers .env)
$ git add .

# Ajoute les fichiers sélectionnés au-dessus au repo dans le premier commit
$ git commit -m "Init"

# pour changer le nom de la branche de 'master' à 'main', pas nécessaire
$ git branch -m main

# Ajouter la source (ton repo GitHub)
$ git remote add origin https://github.com/a2br/monRepo.git

# Pousser tes changements en spécifiant la destination ('origin main', ou 'origin master')
$ git push -u origin main

# Fini, maintenant que peux git add, commit & push comme tu veux
```

**Reply by MixtarMaster** [reply]
D'accord! Merci
Dans mon code j'utilise des infos personnels directement en toute lettres (mot de passe). Je compte faire un settings.json ou on pourra mettre ses infos donc ce problème ne sera plus là lorsque je le rendrait publique. Ca ne pause pas de problème avec git?

**Reply by JsonLinesCode** [reply]
Je me permets de répondre en indiquant qu'il suffit d'ajouter le dit fichier au gitignore pour qu'il ne soit pas envoyé sur le repository public.

**Reply by a2br** [reply]
> D'accord! Merci
> Dans mon code j'utilise des infos personnels directement en toute lettres (mot de passe). Je compte faire un settings.json ou on pourra mettre ses infos donc ce problème ne sera plus là lorsque je le rendrait publique. Ca ne pause pas de problème avec git?

Si, @ModernChocolate dit vrai. Je te conseille la librairie npm dotenv pour les secrets comme ça, même si ça marche aussi un fichier JSON.

**Reply by MixtarMaster** [reply]
Encore Merci! Tout est bon maintenant!!
